### DISCUSSION TITLE
Great plugin luuuuv it :)

### DISCUSSION THREAD

**Discussion by LQdbMusic** [shells-dw/loupedeck-totalmix#2] [Ideas]
Using it with my UCX-2, working great, very snappy, much better than Midi-Crap, **fetten Dank !**

I noticed that the onscreen and on device (Loupedeck) display of values, which btw is a **huge PLUS** for me, is a little bit off, e.g. Main Vol. +-2db: UCX2 shows -44db, Loupedeck and Onscreen shows -42db.

Would be great if you could add future support for FX, like EQ or LowCut, use it all the time in my studio. 
Let me know if I can provide any info for you to go that route, happy to help.

Cheers

**Comment by shells-dw** [comment]
Hi buddy, gerne 😃 

> I noticed that the onscreen and on device (Loupedeck) display of values, which btw is a **huge PLUS** for me, is a little bit off, e.g. Main Vol. +-2db: UCX2 shows -44db, Loupedeck and Onscreen shows -42db.
> 
That's strange. The plugin displays the db values TotalMix sends over OSC. There shouldn't be a discrepancy between TotalMix and the Loupedeck plugin (once the display has updated the value sent by TotalMix), as it doesn't calculate the db value. It just shows what's sent. Maybe the UCX-II display and TotalMix are not lined up then?
That's also the reason why it isn't showing the value in real time. With pan and gain I can calculate the value and show it directly while turning the knob, but I couldn't figure out what logarithm RME uses for their volume faders, it's not a log10 or anything close. I've tried a lot to calculate the value so the display would update with turning but I couldn't get anything calculated that worked for the whole range (this goes down to a float value with 8 digits and it isn't even consistent so a lookup/translation table instead of calculating it doesn't work neither), so I decided I just have it a bit delayed but show exactly what TotalMix sends and figured it'll be fine as sending works in real time and one usually goes by the ears for volume and not so much for printed values 🤷‍♂️

> Would be great if you could add future support for FX, like EQ or LowCut, use it all the time in my studio. Let me know if I can provide any info for you to go that route, happy to help.
> 
All channel buttons have the option to at least disable/enable LowCut, EQ, Comp/Exp and Autolevel, Master Channel has the option to Mute FX and I've just added the option to enable/disable Reverb and Echo with [v1.6.0](https://github.com/shells-dw/loupedeck-totalmix/releases/tag/v1.6.0)
I'd love to utilize the dials for setting individual settings for the FX, however I just own and have access to a Fireface UC that has no FX processor, so the FX options are not accessible to me in TotalMix at all. I can't develop something without being able to see what it does.
This whole OSC implementation RME put in TotalMix is very... exhausting to work with if you ask me. As far as I can see, all these FX functions work on a float level (0.0-1.0 with various degrees of digits) and all have different ranges, so without a device that supports FX I can't really do anything there. At least not in a reasonable amount of time anyway. I could try to brute force that with OSC responses (TotalMix can be asked to change and show these values over OSC but I don't have a chance to see what happens in the TotalMix UI, how it reacts, what happens if I change values in the UI, etc.) - and also that's a bit much effort for a spare time coding around thing for me.
So unless I gain access to a RME interface with FX processor that's sadly nothing I'll be able to implement.

Grüße

**Reply by LQdbMusic** [reply]
> All channel buttons have the option to at least disable/enable LowCut, EQ, Comp/Exp and Autolevel, Master Channel has the option to Mute FX

Stupid me :-) was there all the time, that is so awesome, at least for my type of application :-) Just added this to my config, made my day !

>Maybe the UCX-II display and TotalMix are not lined up then?

Strangely enough they both line up just fine. But the plugin seems to be 1 "click" behind, maybe an offset thing ? or some special update thing in MacOS 12.6.2 on M1 ?

Schönen 4. Advent !

**Reply by shells-dw** [reply]
> > All channel buttons have the option to at least disable/enable LowCut, EQ, Comp/Exp and Autolevel, Master Channel has the option to Mute FX
> 
> Stupid me :-) was there all the time, that is so awesome, at least for my type of application :-) Just added this to my config, made my day !
> 
Perfect 😄 
> > Maybe the UCX-II display and TotalMix are not lined up then?
> 
> Strangely enough they both line up just fine. But the plugin seems to be 1 "click" behind, maybe an offset thing ? or some special update thing in MacOS 12.6.2 on M1 ?
> 
It shouldn't be, let's say that 😆  The value is constantly updated in the background (to reflect changes made in TotalMix directly on the Loupedeck, like when a channel is muted - or a fader is moved) and if the value that's currently held by the plugin is different to what's sent by TotalMix, it'll update the value on the Loupedeck display. So even if it would be a click behind, it'll be updated with the next update run shortly after.
Does this discrepancy show as well when you move the fader in TotalMix UI? The corresponding Loupedeck dial should show the value shortly after.
Honestly I've just done basic testing on MacOS, I mainly use the Fireface on Windows, on MacOS I'm using a UAD interface now, so I just connect the Fireface to my Mac for testing this plugin 😉 
I'll look into it, maybe I can reproduce.
> Schönen 4. Advent !
>
Dir auch, danke!

**Reply by LQdbMusic** [reply]
Maybe my amateur description helps: 

when I change the status, e.g. "Dim Main" on the Loupedeck, the status in TotalMix and on the device itself changes instantly. 

If I change via Totalmix the change on the RME device (and TM of course) is instant, but it takes 2-3 sec. for the status on the Loupedeck to change. Solid Dim button gfx comes up. This behaviour is true for button status. 

But for all numeric dial status on the Loupedeck: the display only updates when I do changes on the Loupedeck device itself again, e.g. turn another dial, or switch App-Profile back and forth. After that "manual" update, all is in sync or value correct

**Reply by LQdbMusic** [reply]
V1.6.1 solved that problem  (at least for the Master Vol.) ! Nice one ! Much better update performance, minimal lag (500ms) on MacOs. Sweeeeet.

**Reply by shells-dw** [reply]
> when I change the status, e.g. "Dim Main" on the Loupedeck, the status in TotalMix and on the device itself changes instantly.
> 
yep, sending basically is instantly, as just one specific value is sent and processed by TotalMix.

> If I change via Totalmix the change on the RME device (and TM of course) is instant, but it takes 2-3 sec. for the status on the Loupedeck to change. Solid Dim button gfx comes up. This behaviour is true for button status.
>
Yeah, that's sadly as fast as I can make it for now. The OSC implementation isn't acting like an API that you could just query, asking for specific values and get back a specific response - or even a full dump. Instead with RME's OSC implementation in TotalMix everything is always "context"-based (but it might be unkown (depending on what's requested) what the current context is based on the reply of the device). It is seperated in different busses (as they call it), namely Input, Playback/Software and Output. That holds some values for all channels in that bus (roughly 192 values on 16-channel setups, on 48 channels it's going up to 560-ish iirc, that per bus, so 3x). That's only "main" functions/values however. Mute, solo, volume, phantom, pan, gain. Everything else a channel can do is on a per-channel basis. So, if you like to set phase invert on input channel 4 for example, the plugin must first make sure it's on the input bus (as it can't really know which bus is active in TotalMix, it needs to put the device to the input channel bus), then it needs to set the currently active channel to channel 4, then it can send the command for phase invert.
So to get info about any main value TotalMix must be asked for a dump of this bus setting, then TotalMix sends the dump (depending on your channel count up to 560-ish values) line per line back, the plugin needs to catch that, split it up and put it somewhere it can programmatically work with the key-value combination, then it needs to do this 2 times more for the other 2 busses and finally a fourth time for the global stuff (which holds 68 values).
This obviously takes time. The plugin needs a bit of time to send the request, TotalMix needs a bit of time to collect and send it (OSC is a network protocol so there is overhead as well even if it's just on the local machine), the plugin has to receive it basically line by line and put it in the right variables, check if there are changes and if so, add the changed values to the function that alerts Loupedeck about the state change, then do that again another 4 times, after that, start right over. So on 16 channels that's roughly 560 values (and on 48 some 1800 iirc) that need to be requested, collected, sent, processed and triggered on changes - which just takes a bit of time.
And that's why I support mirroring only on those main functions. If the plugin would need to query another 62 values for 3x16 channels it would just be taking 30s to update anything.

TL;DR: it's complicated with TotalMix and OSC.

**Reply by shells-dw** [reply]
> V1.6.1 solved that problem (at least for the Master Vol.) ! Nice one ! Much better update performance, minimal lag (500ms) on MacOs. Sweeeeet.

It should actually work on the others already, Master Volume was the only one having that bug - just forgot I wanted to fix that, so thanks for finding it 😆 

If volume dials for input/playback/software aren't working, remove them from the Loupedeck and replace them. Then monitor. It works fine for me on my M1 mac and TotalMixFX 1.78 (which should be the newest version)

**Reply by LQdbMusic** [reply]
> thanks for finding it 😆

lol. ur welcome

> it's complicated with TotalMix and OSC.

Yeah for sure, I watched that Japanese YT-video on TotalMix Midi/OSC implementation and gave up after he started talking about that bussing thing :-) Flexible for all possible channel counts and devices for sure, but hard to wrap your head around IMHO.
The RME forums and manuals are not really that helpful in this regard. But anyhow: for me the speed with 1.6.1 is just fine.

> If volume dials for input/playback/software aren't working, remove them from the Loupedeck and replace them.

I just did but with no luck. The Main Vol. works fine by now, the other still got that weird behaviour. I am also on TM 1.78

**Reply by shells-dw** [reply]
> > it's complicated with TotalMix and OSC.
> 
> Yeah for sure, I watched that Japanese YT-video on TotalMix Midi/OSC implementation and gave up after he started talking about that bussing thing :-) Flexible for all possible channel counts and devices for sure, but hard to wrap your head around IMHO. The RME forums and manuals are not really that helpful in this regard. But anyhow: for me the speed with 1.6.1 is just fine.
> 
feels like they just implemented some sort of MIDI over ethernet and used something that roughly looks like OSC (because there are libs available that handle OSC and it's nothing they'd have had to develop from scratch) but acts like MIDI "under the hood", so they didn't have to implement a proper OSC implementation - but that's just guessing on my part...
OSC at least is somewhat documented, MIDI is a whole different mess. I've supported it in my StreamDeck version of the plugin years ago, but decided just not to bother with it for the Loupedeck rewrite.

> > If volume dials for input/playback/software aren't working, remove them from the Loupedeck and replace them.
> 
> I just did but with no luck. The Main Vol. works fine by now, the other still got that weird behaviour. I am also on TM 1.78
>
Huh. 🤔 
I can't see a reason why. It shouldn't make a difference if it runs on MacOS or Windows (ignoring that it works for me on both) as it's using the same functions as other parts that work for you. Unless the bus dump TotalMix sends via OSC looks different or doesn't contain the values for the volume (which it shouldn't or it wouldn't update on dial turns either), TotalMix should always act the same, just different devices support different functions - meaning for example I can enable reverb through OSC and TotalMix will report it back enabled - it just doesn't do anything as my interface doesn't have that). I'll think about how to log/approach that to figure out what's going on.
Well. Since it sounds like you worked yourself in a bit too, would you be able to send me what your TotalMix dumps for `/1/busInput`?

**Reply by LQdbMusic** [reply]
I checked it with OSCulator on Port 9001, this is what happens when I turn Headphone Vol. (Volume 5 on my device) knob: 

![Bildschirmfoto 2022-12-18 um 21 39 44](https://user-images.githubusercontent.com/120818622/208318453-a751ea12-4ecf-4b72-8e36-23a68f847805.png)

Super snappy in realtime from 0 to 1 IMHO, /1/busInput stays "low" at 0. 
If I turn up e.g. Input-Volume Analog 1-2, it goes "high" to 1, so switch busses afaik correctly.

**Reply by shells-dw** [reply]
/1/busInput will trigger the device to respond with a full dump of the input bus values, and it's actually OSC bundles not only messages (which doesn't make sense to bundle a single message in a bundle, but that's how it's coming out of TotalMix), so your app probably doesn't parse it right. Should be a list of all sorts of settings.
Yeah fair enough, I'll look into it tomorrow and see if I can find anything.
I'll keep you posted.

**Reply by shells-dw** [reply]
I've opened an [issue](https://github.com/shells-dw/loupedeck-totalmix/issues/4) for the issue with your volume dials, let's discuss this specific issue there, here it just gets buried
